MTKA-1334: Add opt-in usage tracking to the plugin.

matthewguywright opened the pull request:
## Description

Adds opt-in usage tracking to the plugin.

At this time, this records an event when:
- A new post is published
- A model is created, updated, or deleted
- A field is created, updated, or deleted
- The model list is viewed

Todo:
- [x] Verify events are properly recorded
- [x] Ensure code/text naming is consistent and appropriate
- [x] Tests

matthewguywright committed:
scripts in place

matthewguywright committed:
ga shared file

matthewguywright committed:
updated per PR

matthewguywright committed:
updated script for gtag

matthewguywright committed:
Update shared_assets.php

matthewguywright committed:
anonymize ip

matthewguywright committed:
Merge branch 'main' into feature/mw-bh-1156-1157

matthewguywright committed:
Merge branch 'main' into feature/mw-bh-1156-1157

matthewguywright committed:
removed old gtags

matthewguywright committed:
init commit with ga-4-react

matthewguywright committed:
Revert "init commit with ga-4-react"

This reverts commit b3abf6e532ce4285adbfb9c5d0afe88422acaf7d.

matthewguywright committed:
init commit with react-ga4

matthewguywright committed:
updated to include ip anonymization

matthewguywright committed:
cleanup for usage

matthewguywright committed:
updated base level GA4 in place

matthewguywright committed:
per PR

matthewguywright committed:
Update App.jsx

matthewguywright committed:
Update App.jsx

nickcernis committed:
[NOJIRA] Initialise Analytics (#251)

* chore: ensure return value is always an array (#248)

`get_acm_taxonomies()` returns the value from `get_option()`, which can be filtered by other code. Casting the return value to array ensures our code doesn't fatal error due to an invalid return type caused by other code.

* feat: add simple analytics helpers

* chore: alias analytics.js to acm-analytics

So that we don't need relative paths in our imports.

* chore: initialize analytics in the publisher app

We can also send the pageview here, because there are no separate
routes in the publisher app.

This should eventually be gated by a check for user opt-in.

* chore: initialize analytics in the settings app

We can't send the pageview at the App root level here, because every
view uses this App as the wrapper, so a visit to the taxonomies page
will trigger “ACM Models Home” incorrectly.

We can trigger the pageview from the specific view instead.

This should eventually be gated by a check for user opt-in.

* chore: send page view when visiting models list

* fix: adjust initialization logic

So that initialization always occurs before an event is triggered.

Co-authored-by: John Parris <john.parris@wpengine.com>

matthewguywright committed:
model events

matthewguywright committed:
deletion

matthewguywright committed:
fields events

matthewguywright committed:
updated per PR

matthewguywright committed:
Merge branch 'main' into feature/google-analytics-4

matthewguywright committed:
Update includes/settings/js/src/components/EditModelModal.jsx

Co-authored-by: John Parris <john.parris@wpengine.com>

ChrisWiegman reviewed (changes requested):
The generic events are great. What worries me here is the disclosure of "what" they're building (see screenshots below). We may need to kill pageviews. We should never know what they call something, just that they're building something.

![Screen Shot 2021-08-27 at 10 29 36](https://user-images.githubusercontent.com/394675/131146095-46f00ee0-7b89-44d7-9503-d60c5895c616.png)
![Screen Shot 2021-08-27 at 10 29 46](https://user-images.githubusercontent.com/394675/131146101-abe1f05c-ea79-4cf2-8ad7-690692cd1e9c.png)
![Screen Shot 2021-08-27 at 10 30 24](https://user-images.githubusercontent.com/394675/131146104-d7c2d30a-9637-4af5-939e-50076d23007c.png)

matthewguywright committed:
updated per PR

matthewguywright committed:
Merge branch 'main' into feature/google-analytics-4

matthewguywright committed:
Update DeleteModelModal.jsx

matthewguywright committed:
added useState

matthewguywright committed:
settings page placeholder

mindctrl committed:
feat: register plugin settings, sections, and fields.

mindctrl committed:
chore: add `wp-core-data` script dependency

mindctrl committed:
chore: pass usage tracking setting to `wp_localize_script`

mindctrl committed:
chore: move `add_action` adjacent to callback function

mindctrl committed:
chore: save settings to database when radio option toggled

matthewguywright committed:
Merge branch 'main' into feature/mw-metadata

mindctrl committed:
chore: remove unused code

mindctrl committed:
chore: remove unused import

mindctrl committed:
chore: add "acm" function prefix to phpcs config

mindctrl committed:
chore: add `acm_usage_tracking_enabled` function.

This function determines if opt-in usage tracking is enabled.

mindctrl committed:
chore: add query arg to location URL when opt-in usage tracking is enabled.

mindctrl committed:
Merge branch 'main' into feature/google-analytics-4

mindctrl committed:
chore: fix issues caused by improper merge conflict resolution

mindctrl committed:
chore: add Parcel alias for `@wordpress/data`

mindctrl committed:
chore: import `dispatch` from `@wordpress/data`

mindctrl committed:
chore: add npm deps that get aliased out of build files

mindctrl committed:
chore: pass usage tracking setting to `wp_localize_script`

mindctrl committed:
chore: use helper function to get usage tracking value

mindctrl committed:
wip: send event when post published

mindctrl committed:
chore: designate react-ga4 as production dependency

mindctrl committed:
chore: change event category to "post" [skip ci]

mindctrl committed:
chore: remove unused import

mindctrl committed:
chore: send event instead of pageview when viewing model list. [skip ci]

Only send the event when opt-in usage tracking is enabled.

mindctrl committed:
chore: remove unused import [skip ci]

mindctrl committed:
chore: remove unused imports

mindctrl committed:
chore: remove unused `useState` calls

mindctrl committed:
chore: send event when model created if usage tracking enabled

mindctrl committed:
chore: send event when model deleted if usage tracking enabled

mindctrl committed:
chore: send event when model updated if usage tracking enabled

mindctrl committed:
chore: move `view_models_list` event to proper component

mindctrl committed:
chore: remove unused import

mindctrl committed:
chore: send event when a field is removed from a model

mindctrl committed:
chore: send event when a field is created or updated

mindctrl commented on includes/settings/js/src/components/DeleteModelModal.jsx:
```diff
@@ -118,6 +120,12 @@ export function DeleteModelModal({ modalIsOpen, setModalIsOpen, model }) {
 					await deleteModel(slug)
 						.then((res) => {
 							if (res.success) {
+								if (atlasContentModeler.usageTrackingEnabled) {
```
I don't love adding this kind of logic everywhere we want an event. An alternative is to put the logic inside the functions that send the events. However, the `usageTrackingEnabled` bit that is passed through `wp_localize_script` is done for two separate scripts (publisher and settings) and exposed through two different global variables by WP. We don't have a single shared script that is enqueued and localized in both places. Something to consider for the near future.

nickcernis replied:
@mindctrl Perhaps something like this for now?

```diff
diff --git a/includes/shared-assets/js/analytics.js b/includes/shared-assets/js/analytics.js
index dfa17c6..7332f8b 100644
--- a/includes/shared-assets/js/analytics.js
+++ b/includes/shared-assets/js/analytics.js
@@ -7,6 +7,10 @@ import ReactGA from "react-ga4";
 
 const ANALYTICS_ID = "G-S056CLLZ34";
 
+const telemetryEnabled =
+	atlasContentModelerFormEditingExperience?.usageTrackingEnabled ||
+	atlasContentModeler?.usageTrackingEnabled;
+
 const maybeInitializeAnalytics = () => {
 	if (!ReactGA?.isInitialized) {
 		ReactGA.initialize(ANALYTICS_ID, {
@@ -17,13 +21,18 @@ const maybeInitializeAnalytics = () => {
 
 export const sendEvent = (data) => {
 	maybeInitializeAnalytics();
-	return ReactGA.event(data);
+	if (telemetryEnabled) {
+		return ReactGA.event(data);
+	}
 };
 
 export const sendPageView = (page = "") => {
 	maybeInitializeAnalytics();
-	return ReactGA.send({
-		hitType: "pageview",
-		page,
-	});
+
+	if (telemetryEnabled) {
+		return ReactGA.send({
+			hitType: "pageview",
+			page,
+		});
+	}
 };
```

mindctrl commented:
This is ready for a preliminary review. I updated the original post with the remaining tasks.

nickcernis commented on includes/settings/js/src/components/Settings.jsx:
```diff
@@ -0,0 +1,96 @@
+import { __ } from "@wordpress/i18n";
+import { dispatch } from "@wordpress/data";
+
+export default function Settings() {
```
This isn't saving settings quite right for me at the moment:

- Disabled isn't active by default (i.e. when `atlas_content_modeler_usage_tracking` is empty).
- Clicking 'disabled' then refreshing shows no option as active.

This change seems to fix both issues:

```diff
diff --git a/includes/settings/js/src/components/Settings.jsx b/includes/settings/js/src/components/Settings.jsx
index bdaff70..77635d6 100644
--- a/includes/settings/js/src/components/Settings.jsx
+++ b/includes/settings/js/src/components/Settings.jsx
@@ -8,12 +8,12 @@ export default function Settings() {
 		atlasContentModeler.usageTrackingEnabled
 	);
 
-	function saveUsageTrackingSetting(event) {
+	function saveUsageTrackingSetting(value) {
 		// @todo catch save errors
 		dispatch("core").saveSite({
-			atlas_content_modeler_usage_tracking: event.target.value,
+			atlas_content_modeler_usage_tracking: value,
 		});
-		setUsageTracking(event.target.value);
+		setUsageTracking(value);
 	}
 
 	return (
@@ -51,10 +51,13 @@ export default function Settings() {
 													name="atlas-content-modeler-settings[usageTracking]"
 													value="0"
 													checked={
-														usageTracking === "0"
+														usageTracking === "0" ||
+														!usageTracking
 													}
-													onChange={
-														saveUsageTrackingSetting
+													onChange={() =>
+														saveUsageTrackingSetting(
+															"0"
+														)
 													}
 												></input>
 												{__(
@@ -70,8 +73,10 @@ export default function Settings() {
 												name="atlas-content-modeler-settings[usageTracking]"
 												value="1"
 												checked={usageTracking === "1"}
-												onChange={
-													saveUsageTrackingSetting
+												onChange={() =>
+													saveUsageTrackingSetting(
+														"1"
+													)
 												}
 											></input>
 											<label
```

nickcernis commented on includes/settings/js/src/components/Settings.jsx:
```diff
@@ -0,0 +1,96 @@
+													id="atlas-content-modeler-settings[usageTrackingDisabled]"
+													name="atlas-content-modeler-settings[usageTracking]"
+													value="0"
+													checked={
+														usageTracking === "0"
+													}
```
This isn't saving settings quite right for me at the moment:

- Disabled isn't active by default (i.e. when atlas_content_modeler_usage_tracking is empty).
- Clicking 'disabled' then refreshing shows no option as active.

This change seems to fix both issues:

```suggestion
													checked={
														usageTracking === "0" ||
														!usageTracking
													}
```

mindctrl committed:
test: confirm `wp-core-data` script loads on settings page

mindctrl committed:
test: confirm all external script dependencies load on settings page

mindctrl committed:
test: confirm settings registration callbacks are attached

mindctrl committed:
test: confirm usage tracking helper function

mindctrl committed:
fix: default checked state when option is empty

Co-authored-by: Nick Cernis <nick@cern.is>

mindctrl committed:
chore: simplify opt-in check logic

mindctrl committed:
chore: remove unnecessary setup and teardown methods

mindctrl committed:
test: confirm that publisher callbacks are attached to hooks

mindctrl committed:
chore: add test method for visiting plugin settings page

mindctrl committed:
test: confirm plugin settings page functionality

Tests that the plugin setting page renders the
opt-in usage tracking option, and that toggling
the setting saves the value to the database
as expected.

mindctrl committed:
chore: DRY

mindctrl committed:
test: confirm return value of `add_published_query_arg_to_location()`

Tests that `acm-post-published` string is added to the post edit URL
when a post is published and usage tracking is enabled.

Tests that the post edit URL is not modified when a post is published
if usage tracking is disabled.

nickcernis commented on includes/settings/js/src/components/DeleteModelModal.jsx:
```diff
@@ -6,6 +7,7 @@ import { removeSidebarMenuItem } from "../utils";
 import { getRelationships } from "../queries";
 import { showError } from "../toasts";
 import { useHistory } from "react-router-dom";
+import { sendEvent } from "../../../../shared-assets/js/analytics";
```
Worth using the alias here for consistency?

```suggestion
import { sendEvent } from "acm-analytics";
```

mindctrl replied:
Ah yes, I think my editor auto-added the path. Good catch.

nickcernis reviewed (dismissed):
One suggestion above but otherwise looks good! Might also benefit from a changelog entry under "Unreleased"?

mindctrl committed:
chore: replace import path with alias

Co-authored-by: Nick Cernis <nick@cern.is>

mindctrl committed:
chore: add changelog entry

nickcernis commented on CHANGELOG.md:
```diff
@@ -1,5 +1,9 @@
 
+## Unreleased
+### Added
+- Opt-in anonymous usage tracking to help us make Atlas Content Modeler better.
```
Worth adding this for clarification?

```suggestion
- Opt-in anonymous usage tracking to help us make Atlas Content Modeler better (disabled by default).
```

mindctrl replied:
Definitely! Thanks!

mindctrl committed:
chore: clarify changelog entry

Co-authored-by: Nick Cernis <nick@cern.is>

mindctrl committed:
chore: rename events for better legibility

mindctrl committed:
Merge branch 'main' into feature/google-analytics-4

mindctrl committed:
fix: only initialize ReactGA when necessary
